File: JavaCode/src/com/ysx/leetcode/easy/ToeplitzMatrix.java
package com.ysx.leetcode.easy;

/**
 * 766. Toeplitz Matrix
 * https://leetcode.com/problems/toeplitz-matrix/description/
 */
public class ToeplitzMatrix {

    public static void main(String[] args) {
        ToeplitzMatrix test = new ToeplitzMatrix();
        int[][] matrix = {{1, 2, 3, 4}, {5, 1, 2, 3}, {9, 5, 1, 2}};
        System.out.println(test.isToeplitzMatrix(matrix));

        int[][] matrix2 = {{1, 2}, {2, 2}};
        System.out.println(test.isToeplitzMatrix(matrix2));

        int[][] matrix3 = {{65, 98, 57}};
        System.out.println(test.isToeplitzMatrix(matrix3));
    }

    public boolean isToeplitzMatrix(int[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return false;
        }
        int m = matrix.length;
        int n = matrix[0].length;
        for (int i = 0; i < m; i++) {
            int value = matrix[i][0];
            for (int j = 1; j + i < m && j < n; j++) {
                if (matrix[i + j][j] != value) {
                    return false;
                }
            }
        }
        for (int j = 1; j < n; j++) {
            int value = matrix[0][j];
            for (int i = 1; i < m && i + j < n; i++) {
                if (matrix[i][i + j] != value) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: KotlinCode/src/leetcode/easy/ToeplitzMatrix.kt
package leetcode.easy

class ToeplitzMatrix {
    fun isToeplitzMatrix(matrix: Array<IntArray>): Boolean {
        if (matrix.isEmpty() || matrix[0].isEmpty()) {
            return false
        }
        val m = matrix.size
        val n = matrix[0].size
        for (i in 0 until m) {
            val value = matrix[i][0]
            var j = 1
            while (j + i < m && j < n) {
                if (matrix[i + j][j] != value) {
                    return false
                }
                j++
            }
        }
        for (j in 1 until n) {
            val value = matrix[0][j]
            var i = 1
            while (i < m && i + j < n) {
                if (matrix[i][i + j] != value) {
                    return false
                }
                i++
            }
        }
        return true
    }
}

fun main(args: Array<String>) {
    val test = ToeplitzMatrix()
    val matrix = arrayOf(intArrayOf(1, 2, 3, 4), intArrayOf(5, 1, 2, 3), intArrayOf(9, 5, 1, 2))
    println(test.isToeplitzMatrix(matrix))

    val matrix2 = arrayOf(intArrayOf(1, 2), intArrayOf(2, 2))
    println(test.isToeplitzMatrix(matrix2))

    val matrix3 = arrayOf(intArrayOf(65, 98, 57))
    println(test.isToeplitzMatrix(matrix3))
}